signac/core/validators.py

```diff
@@ -0,0 +1,83 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Validators for SyncedCollection API."""
+import warnings
+from collections.abc import Mapping
+from collections.abc import Sequence
+
+from ..errors import InvalidKeyError
+from ..errors import KeyTypeError
+
+try:
+    import numpy
+    NUMPY = True
+except ImportError:
+    NUMPY = False
+
+
+def NoDotInKey(data):
+    """Raise an exception if there is a dot (``.``) in a mapping's key.
+
+    Parameters
+    ----------
+    data
+        Data to validate.
+
+    Raises
+    ------
+    KeyTypeError
+        If keys have unsupported data type.
+    InvalidKeyError
+        If key is not supported.
+    """
+    VALID_KEY_TYPES = (str, int, bool, type(None))
```

Comment:
Validation is done before we pass the data to the backend. So, I think we need to support all the non-string key types here.
> (we aren't going to support non-string keys in new backends)

I think we are only removing support for non-string keys in JSON backend.

Reply:
@bdice @vishav1771 this change is not necessary and the current implementation is in line with the design that I proposed. The `SyncedCollection` layer is to completely support the use cases in signac while remaining entirely independent of signac. Any validator can be registered with any class prior to its use, which that signac can (and should) apply the `JSON_format_validator` to _all_ back ends. However, a `SyncedAttrDict` used outside the context of signac (which is completely feasible) still needs to avoid having dots in keys because that breaks the attribute-based access to stored items, which is why this validator is shipped with the framework and registered by default to this class. In the context of signac, we'll need both validators.

One thing that we haven't discussed but we might need to at some point is, if applying both validators becomes a noticeable performance drag we may need to add a method to clear validators and write a single validator that performs both functions. However, at the moment this would be premature optimization.

Reply:
I came here straight from a comment on Slack and I think I missed the point here. @vishav1771 I agree with @bdice that we shouldn't support non-str keys. I actually think we should remove the automatic casting to str altogether. That's deprecated behavior, and I think when synced collections get merged we should go ahead and break it, with the intention of releasing synced collections as part of the 2.x line of signac.

Reply:
@vyasr Let's drop non-str keys in a follow-up PR. I don't want to introduce any further design changes here right now. This has been through many iterations and I'm confident in the current behavior doing what it intends to do correctly (even though it's not what we want for signac 2.0).

Reply:
OK wow my computer just didn't refresh any of your comments. I made my last two without even seeing the ones immediately before them, but now I think I see the whole thread.

Yes, more precisely _signac_ will apply the JSON format validator to all back ends. I make this distinction to make the point that nowhere in any of the files defining SyncedCollection subclasses will we include `cls.add_validator(JSON_format_validator)` except in collection_json.py. Therefore, whatever back end the user chooses to use with signac, signac will internally need to call cls.add_validator prior to creating any instances of that class.

I'm fine with waiting to that until a future PR. The necessary conversion for backwards compatibility has been added to the JSONCollection class's sync method, so we would just remove that at some future point.

Reply:
Agreed, the validator layer(s) should be independent of the backend and we should apply one consistent set of validators to all backends that we natively support.

Reply:
I agree with @csadorf we need to have one set of validation that runs for every backend. It would be really annoying to be using a redis backend for performance but when it comes time to archive the data, find out that the keys used for redis are incompatible with json. I also agree that we should wait for a future PR to modify the current behavior.

Reply:
*ugh github ate my comment*
tl;dr I agree with @csadorf that we need to have one consistent set of validation criteria. It would be really annoying for a user to use a redis backend for performance but then when it comes time to archive the project, run issues when switching to json. I also agree we can wait for another PR to modify behavior.